File: lib/sidereon/space_weather.ex
defmodule Sidereon.SpaceWeather do
  @moduledoc """
  Space-weather tables for drag and decay inputs.

  The parser and lookup policy are implemented in the core NIF. This module
  reads optional files, owns the Elixir structs, and forwards table operations to
  the parsed native resource.
  """

  alias Sidereon.Drag
  alias Sidereon.NIF

  @enforce_keys [:handle]
  defstruct [:handle]

  @type t :: %__MODULE__{handle: reference()}

  defmodule Sample do
    @moduledoc """
    Space-weather values plus source metadata for one lookup.
    """
    @enforce_keys [:space_weather, :class, :ap_defaulted]
    defstruct [:space_weather, :class, :ap_defaulted]

    @type t :: %__MODULE__{
            space_weather: Drag.SpaceWeather.t(),
            class: :observed | :interpolated | :daily_predicted | :monthly_predicted,
            ap_defaulted: boolean()
          }
  end

  defmodule Policy do
    @moduledoc """
    Lookup policy for rejecting lower-trust rows or monthly Ap defaults.
    """
    defstruct allow_interpolated: true,
              allow_daily_predicted: true,
              allow_monthly_predicted: true,
              require_geomagnetic: false

    @type t :: %__MODULE__{
            allow_interpolated: boolean(),
            allow_daily_predicted: boolean(),
            allow_monthly_predicted: boolean(),
            require_geomagnetic: boolean()
          }
  end

  defmodule Coverage do
    @moduledoc """
    J2000-second coverage bounds for a parsed space-weather table.
    """
    @enforce_keys [:first_j2000_s, :end_j2000_s]
    defstruct [:first_j2000_s, :last_observed_j2000_s, :last_daily_predicted_j2000_s, :end_j2000_s]

    @type t :: %__MODULE__{
            first_j2000_s: float(),
            last_observed_j2000_s: float() | nil,
            last_daily_predicted_j2000_s: float() | nil,
            end_j2000_s: float()
          }
  end

  @type error_reason ::
          :unrecognized_format
          | :not_text
          | {:malformed, non_neg_integer(), String.t()}
          | {:before_coverage, float(), float()}
          | {:after_coverage, float(), float()}
          | {:missing_data, integer(), integer(), integer(), String.t()}
          | {:rejected_by_policy, atom(), integer(), integer(), integer()}
          | {:invalid_epoch, float()}
          | term()

  @spec parse(binary()) :: {:ok, t()} | {:error, error_reason()}
  def parse(bytes) when is_binary(bytes) do
    case NIF.space_weather_parse(bytes) do
      {:ok, handle} when is_reference(handle) -> {:ok, %__MODULE__{handle: handle}}
      {:error, _} = err -> err
      other -> {:error, other}
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @spec load(String.t()) :: {:ok, t()} | {:error, term()}
  def load(path) when is_binary(path) do
    with {:ok, bytes} <- File.read(path), do: parse(bytes)
  end

  @spec space_weather_at(t(), number()) :: {:ok, Drag.SpaceWeather.t()} | {:error, error_reason()}
  def space_weather_at(%__MODULE__{handle: handle}, epoch_j2000_s) when is_number(epoch_j2000_s) do
    case NIF.space_weather_space_weather_at(handle, epoch_j2000_s / 1.0) do
      {:ok, fields} -> {:ok, to_space_weather(fields)}
      {:error, _} = err -> err
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @spec sample_at(t(), number()) :: {:ok, Sample.t()} | {:error, error_reason()}
  def sample_at(%__MODULE__{handle: handle}, epoch_j2000_s) when is_number(epoch_j2000_s) do
    case NIF.space_weather_sample_at(handle, epoch_j2000_s / 1.0) do
      {:ok, fields} -> {:ok, to_sample(fields)}
      {:error, _} = err -> err
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @spec sample_at_with_policy(t(), number(), Policy.t() | keyword()) :: {:ok, Sample.t()} | {:error, error_reason()}
  def sample_at_with_policy(%__MODULE__{handle: handle}, epoch_j2000_s, policy) when is_number(epoch_j2000_s) do
    case NIF.space_weather_sample_at_with_policy(handle, epoch_j2000_s / 1.0, policy_map(policy)) do
      {:ok, fields} -> {:ok, to_sample(fields)}
      {:error, _} = err -> err
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @spec ap_array_at(t(), number()) :: {:ok, [float()]} | {:error, error_reason()}
  def ap_array_at(%__MODULE__{handle: handle}, epoch_j2000_s) when is_number(epoch_j2000_s) do
    NIF.space_weather_ap_array_at(handle, epoch_j2000_s / 1.0)
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @spec coverage(t()) :: {:ok, Coverage.t()} | {:error, term()}
  def coverage(%__MODULE__{handle: handle}) do
    case NIF.space_weather_coverage(handle) do
      {:ok, fields} -> {:ok, struct!(Coverage, fields)}
      {:error, _} = err -> err
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @spec to_csv_text(t()) :: {:ok, String.t()} | {:error, term()}
  def to_csv_text(%__MODULE__{handle: handle}), do: {:ok, NIF.space_weather_to_csv_text(handle)}

  @spec to_txt_text(t()) :: {:ok, String.t()} | {:error, term()}
  def to_txt_text(%__MODULE__{handle: handle}), do: {:ok, NIF.space_weather_to_txt_text(handle)}

  def parse!(bytes), do: bang(parse(bytes))
  def load!(path), do: bang(load(path))
  def space_weather_at!(table, epoch_j2000_s), do: bang(space_weather_at(table, epoch_j2000_s))
  def sample_at!(table, epoch_j2000_s), do: bang(sample_at(table, epoch_j2000_s))

  def sample_at_with_policy!(table, epoch_j2000_s, policy),
    do: bang(sample_at_with_policy(table, epoch_j2000_s, policy))

  def ap_array_at!(table, epoch_j2000_s), do: bang(ap_array_at(table, epoch_j2000_s))
  def coverage!(table), do: bang(coverage(table))

  defp to_sample(fields) do
    %Sample{
      space_weather: to_space_weather(fields.space_weather),
      class: fields.class,
      ap_defaulted: fields.ap_defaulted
    }
  end

  defp to_space_weather(fields) do
    %Drag.SpaceWeather{f107: fields.f107, f107a: fields.f107a, ap: fields.ap}
  end

  defp policy_map(%Policy{} = policy), do: Map.from_struct(policy)

  defp policy_map(opts) when is_list(opts) do
    defaults = %Policy{}
    struct!(Policy, Keyword.take(opts, Map.keys(Map.from_struct(defaults)))) |> Map.from_struct()
  end

  defp bang({:ok, value}), do: value
  defp bang({:error, reason}), do: raise(ArgumentError, "space-weather operation failed: #{inspect(reason)}")
end

File: native/sidereon_nif/src/space_weather.rs
//! Rustler boundary for core space-weather tables.
//!
//! Parsing, lookup, policy handling, and formatting stay in `sidereon-core`.
//! This module only moves terms across the BEAM boundary and keeps parsed
//! tables in resource handles.

use std::sync::Arc;

use rustler::{Binary, Encoder, Env, ResourceArc, Term};
use sidereon_core::astro::forces::SpaceWeather;
use sidereon_core::astro::forces::SpaceWeatherSource;
use sidereon_core::astro::propagator::decay::estimate_decay_with_source;
use sidereon_core::astro::space_weather::{
    encode_csv, encode_txt, parse, ObservationClass, SpaceWeatherCoverage, SpaceWeatherError,
    SpaceWeatherPolicy, SpaceWeatherSample, SpaceWeatherTable,
};

use crate::drag::{
    decay_config_from_terms, decode_decay_error, DecayEstimateTerm, DragParametersTerm,
};

pub struct SpaceWeatherTableResource {
    pub table: Arc<SpaceWeatherTable>,
}

#[rustler::resource_impl]
impl rustler::Resource for SpaceWeatherTableResource {}

mod atoms {
    rustler::atoms! {
        ok,
        error,
        observed,
        interpolated,
        daily_predicted,
        monthly_predicted,
        unrecognized_format,
        malformed,
        not_text,
        before_coverage,
        after_coverage,
        missing_data,
        rejected_by_policy,
        invalid_epoch,
    }
}

#[derive(Debug, Clone, rustler::NifMap)]
pub(crate) struct SpaceWeatherTerm {
    f107: f64,
    f107a: f64,
    ap: f64,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct SpaceWeatherSampleTerm {
    space_weather: SpaceWeatherTerm,
    class: rustler::Atom,
    ap_defaulted: bool,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct SpaceWeatherPolicyTerm {
    allow_interpolated: bool,
    allow_daily_predicted: bool,
    allow_monthly_predicted: bool,
    require_geomagnetic: bool,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct SpaceWeatherCoverageTerm {
    first_j2000_s: f64,
    last_observed_j2000_s: Option<f64>,
    last_daily_predicted_j2000_s: Option<f64>,
    end_j2000_s: f64,
}

impl From<SpaceWeather> for SpaceWeatherTerm {
    fn from(space_weather: SpaceWeather) -> Self {
        Self {
            f107: space_weather.f107,
            f107a: space_weather.f107a,
            ap: space_weather.ap,
        }
    }
}

impl From<SpaceWeatherPolicyTerm> for SpaceWeatherPolicy {
    fn from(policy: SpaceWeatherPolicyTerm) -> Self {
        Self {
            allow_interpolated: policy.allow_interpolated,
            allow_daily_predicted: policy.allow_daily_predicted,
            allow_monthly_predicted: policy.allow_monthly_predicted,
            require_geomagnetic: policy.require_geomagnetic,
        }
    }
}

impl From<SpaceWeatherSample> for SpaceWeatherSampleTerm {
    fn from(sample: SpaceWeatherSample) -> Self {
        Self {
            space_weather: sample.space_weather.into(),
            class: class_atom(sample.class),
            ap_defaulted: sample.ap_defaulted,
        }
    }
}

impl From<SpaceWeatherCoverage> for SpaceWeatherCoverageTerm {
    fn from(coverage: SpaceWeatherCoverage) -> Self {
        Self {
            first_j2000_s: coverage.first_j2000_s,
            last_observed_j2000_s: coverage.last_observed_j2000_s,
            last_daily_predicted_j2000_s: coverage.last_daily_predicted_j2000_s,
            end_j2000_s: coverage.end_j2000_s,
        }
    }
}

fn class_atom(class: ObservationClass) -> rustler::Atom {
    match class {
        ObservationClass::Observed => atoms::observed(),
        ObservationClass::Interpolated => atoms::interpolated(),
        ObservationClass::DailyPredicted => atoms::daily_predicted(),
        ObservationClass::MonthlyPredicted => atoms::monthly_predicted(),
    }
}

fn encode_error<'a>(env: Env<'a>, error: SpaceWeatherError) -> Term<'a> {
    match error {
        SpaceWeatherError::UnrecognizedFormat => {
            (atoms::error(), atoms::unrecognized_format()).encode(env)
        }
        SpaceWeatherError::Malformed { line, reason } => {
            (atoms::error(), (atoms::malformed(), line, reason)).encode(env)
        }
        SpaceWeatherError::NotText => (atoms::error(), atoms::not_text()).encode(env),
        SpaceWeatherError::BeforeCoverage {
            requested_j2000_s,
            first_j2000_s,
        } => (
            atoms::error(),
            (atoms::before_coverage(), requested_j2000_s, first_j2000_s),
        )
            .encode(env),
        SpaceWeatherError::AfterCoverage {
            requested_j2000_s,
            end_j2000_s,
        } => (
            atoms::error(),
            (atoms::after_coverage(), requested_j2000_s, end_j2000_s),
        )
            .encode(env),
        SpaceWeatherError::MissingData {
            year,
            month,
            day,
            field,
        } => (
            atoms::error(),
            (
                atoms::missing_data(),
                year,
                i64::from(month),
                i64::from(day),
                field,
            ),
        )
            .encode(env),
        SpaceWeatherError::RejectedByPolicy {
            class,
            year,
            month,
            day,
        } => (
            atoms::error(),
            (
                atoms::rejected_by_policy(),
                class_atom(class),
                year,
                i64::from(month),
                i64::from(day),
            ),
        )
            .encode(env),
        SpaceWeatherError::InvalidEpoch { epoch_j2000_s_bits } => (
            atoms::error(),
            (atoms::invalid_epoch(), f64::from_bits(epoch_j2000_s_bits)),
        )
            .encode(env),
    }
}

fn encode_lookup<'a, T, F>(
    env: Env<'a>,
    result: Result<T, SpaceWeatherError>,
    encode_ok: F,
) -> Term<'a>
where
    F: FnOnce(Env<'a>, T) -> Term<'a>,
{
    match result {
        Ok(value) => (atoms::ok(), encode_ok(env, value)).encode(env),
        Err(error) => encode_error(env, error),
    }
}

#[rustler::nif(schedule = "DirtyCpu")]
fn space_weather_parse<'a>(env: Env<'a>, bytes: Binary<'a>) -> Term<'a> {
    match parse(bytes.as_slice()) {
        Ok(parsed) => {
            let (table, _diagnostics) = parsed.into_parts();
            (
                atoms::ok(),
                ResourceArc::new(SpaceWeatherTableResource {
                    table: Arc::new(table),
                }),
            )
                .encode(env)
        }
        Err(error) => encode_error(env, error),
    }
}

#[rustler::nif]
fn space_weather_space_weather_at<'a>(
    env: Env<'a>,
    handle: ResourceArc<SpaceWeatherTableResource>,
    epoch_j2000_s: f64,
) -> Term<'a> {
    encode_lookup(
        env,
        handle.table.space_weather_at(epoch_j2000_s),
        |env, sw| SpaceWeatherTerm::from(sw).encode(env),
    )
}

#[rustler::nif]
fn space_weather_sample_at<'a>(
    env: Env<'a>,
    handle: ResourceArc<SpaceWeatherTableResource>,
    epoch_j2000_s: f64,
) -> Term<'a> {
    encode_lookup(env, handle.table.sample_at(epoch_j2000_s), |env, sample| {
        SpaceWeatherSampleTerm::from(sample).encode(env)
    })
}

#[rustler::nif]
fn space_weather_sample_at_with_policy<'a>(
    env: Env<'a>,
    handle: ResourceArc<SpaceWeatherTableResource>,
    epoch_j2000_s: f64,
    policy: SpaceWeatherPolicyTerm,
) -> Term<'a> {
    encode_lookup(
        env,
        handle
            .table
            .sample_at_with_policy(epoch_j2000_s, policy.into()),
        |env, sample| SpaceWeatherSampleTerm::from(sample).encode(env),
    )
}

#[rustler::nif]
fn space_weather_ap_array_at<'a>(
    env: Env<'a>,
    handle: ResourceArc<SpaceWeatherTableResource>,
    epoch_j2000_s: f64,
) -> Term<'a> {
    encode_lookup(env, handle.table.ap_array_at(epoch_j2000_s), |env, ap| {
        ap.to_vec().encode(env)
    })
}

#[rustler::nif]
fn space_weather_coverage<'a>(
    env: Env<'a>,
    handle: ResourceArc<SpaceWeatherTableResource>,
) -> Term<'a> {
    (
        atoms::ok(),
        SpaceWeatherCoverageTerm::from(handle.table.coverage()),
    )
        .encode(env)
}

#[rustler::nif(schedule = "DirtyCpu")]
fn space_weather_to_csv_text(handle: ResourceArc<SpaceWeatherTableResource>) -> String {
    encode_csv(&handle.table)
}

#[rustler::nif(schedule = "DirtyCpu")]
fn space_weather_to_txt_text(handle: ResourceArc<SpaceWeatherTableResource>) -> String {
    encode_txt(&handle.table)
}

#[rustler::nif(schedule = "DirtyCpu")]
#[allow(clippy::too_many_arguments)]
fn drag_estimate_decay_with_space_weather_table<'a>(
    env: Env<'a>,
    state: crate::drag::CartesianStateTerm,
    params: DragParametersTerm,
    force_model_name: String,
    abs_tol: f64,
    rel_tol: f64,
    reentry_altitude_km: f64,
    scan_step_s: f64,
    crossing_tolerance_s: f64,
    max_duration_s: f64,
    max_scan_samples: u32,
    table: ResourceArc<SpaceWeatherTableResource>,
) -> rustler::NifResult<Term<'a>> {
    let mut config = decay_config_from_terms(
        params,
        force_model_name,
        abs_tol,
        rel_tol,
        reentry_altitude_km,
        scan_step_s,
        crossing_tolerance_s,
        max_duration_s,
        max_scan_samples,
    )?;
    config.mu_km3_s2 = None;
    let source = SpaceWeatherSource::Table(table.table.clone());

    Ok(
        match estimate_decay_with_source(crate::drag::state_from_term(state), &config, &source) {
            Ok(estimate) => (
                atoms::ok(),
                DecayEstimateTerm {
                    time_to_decay_s: estimate.time_to_decay_s,
                    reentry_state: crate::drag::state_to_term(estimate.reentry_state),
                    reentry_altitude_km: estimate.reentry_altitude_km,
                },
            )
                .encode(env),
            Err(error) => decode_decay_error(env, error),
        },
    )
}
